File: Week7/src/controllers/productController.js
const Product = require("../models/Product");

// GET /api/products
const getAllProducts = async (req, res) => {
    try {
        const products = await Product.find();
        res.status(200).json(products);
    } catch (error) {
        res.status(500).json({ message: "Lỗi server", error: error.message });
    }
};

// GET /api/products/:id
const getProductById = async (req, res) => {
    try {
        const product = await Product.findById(req.params.id);
        if (!product) return res.status(404).json({ message: "Không tìm thấy sản phẩm" });
        res.status(200).json(product);
    } catch (error) {
        res.status(500).json({ message: "Lỗi server", error: error.message });
    }
};

// POST /api/products
const createProduct = async (req, res) => {
    try {
        const newProduct = new Product(req.body);
        const saved = await newProduct.save();
        res.status(201).json(saved);
    } catch (error) {
        if (error.name === "ValidationError")
            return res.status(400).json({ message: error.message });
        res.status(500).json({ message: "Lỗi server", error: error.message });
    }
};

// PUT /api/products/:id
const updateProduct = async (req, res) => {
    try {
        const updated = await Product.findByIdAndUpdate(
            req.params.id,
            req.body,
            { new: true, runValidators: true }
        );
        if (!updated) return res.status(404).json({ message: "Không tìm thấy sản phẩm" });
        res.status(200).json(updated);
    } catch (error) {
        if (error.name === "ValidationError")
            return res.status(400).json({ message: error.message });
        res.status(500).json({ message: "Lỗi server", error: error.message });
    }
};

// DELETE /api/products/:id
const deleteProduct = async (req, res) => {
    try {
        const deleted = await Product.findByIdAndDelete(req.params.id);
        if (!deleted) return res.status(404).json({ message: "Không tìm thấy sản phẩm" });
        res.status(200).json({ message: "Xóa thành công" });
    } catch (error) {
        res.status(500).json({ message: "Lỗi server", error: error.message });
    }
};

module.exports = { getAllProducts, getProductById, createProduct, updateProduct, deleteProduct };

File: Week7/app/controllers/product_controller.py
from flask import request, jsonify
from flask_pymongo import PyMongo
from bson import ObjectId
from bson.errors import InvalidId
from datetime import datetime
from app import mongo

# GET /api/products


def get_all_products():
    products = mongo.db.products.find()
    return jsonify([product_serializer(p) for p in products]), 200

# GET /api/products/<id>


def get_product_by_id(id):
    try:
        product = mongo.db.products.find_one({"_id": ObjectId(id)})
        if not product:
            return jsonify({"message": "Không tìm thấy sản phẩm"}), 404
        return jsonify(product_serializer(product)), 200
    except InvalidId:
        return jsonify({"message": "ID không hợp lệ"}), 400

# POST /api/products


def create_product():
    data = request.get_json()
    if not data.get("name") or not data.get("price"):
        return jsonify({"message": "Thiếu name hoặc price"}), 400

    new_product = {
        "name": data["name"],
        "description": data.get("description", ""),
        "price": data["price"],
        "stock": data.get("stock", 0),
        "createdAt": datetime.utcnow(),
        "updatedAt": datetime.utcnow(),
    }
    result = mongo.db.products.insert_one(new_product)
    new_product["_id"] = result.inserted_id
    return jsonify(product_serializer(new_product)), 201

# PUT /api/products/<id>


def update_product(id):
    try:
        data = request.get_json()
        updated_fields = {
            "name": data.get("name"),
            "description": data.get("description"),
            "price": data.get("price"),
            "stock": data.get("stock"),
            "updatedAt": datetime.utcnow(),
        }
        # Bỏ những field None
        updated_fields = {k: v for k,
                          v in updated_fields.items() if v is not None}

        result = mongo.db.products.find_one_and_update(
            {"_id": ObjectId(id)},
            {"$set": updated_fields},
            return_document=True
        )
        if not result:
            return jsonify({"message": "Không tìm thấy sản phẩm"}), 404
        return jsonify(product_serializer(result)), 200
    except InvalidId:
        return jsonify({"message": "ID không hợp lệ"}), 400

# DELETE /api/products/<id>


def delete_product(id):
    try:
        result = mongo.db.products.find_one_and_delete({"_id": ObjectId(id)})
        if not result:
            return jsonify({"message": "Không tìm thấy sản phẩm"}), 404
        return jsonify({"message": "Xóa thành công"}), 200
    except InvalidId:
        return jsonify({"message": "ID không hợp lệ"}), 400


def product_serializer(product) -> dict:
    return {
        "_id": str(product["_id"]),
        "name": product.get("name", ""),
        "description": product.get("description", ""),
        "price": product.get("price", 0),
        "stock": product.get("stock", 0),
        "createdAt": str(product.get("createdAt", "")),
        "updatedAt": str(product.get("updatedAt", "")),
    }
